Automattic/wp-calypso
# Activity Log: Add the notion of retention limits + an upsell for them

Resolves `1200412004370260-as-1200532973279678`.

#### Changes proposed in this Pull Request

* Create a new feature flag `activity-log/retention-policies` to enable/disable Activity Log retention policies (disabled by default for now).
* Create a new set of reducers, actions, etc. for storing and retrieving a site's Activity Log retention policy.
* Create a new component `RetentionLimitUpsell` to offer customers the ability to upgrade their Activity Log retention policy, visible when they reach the oldest Activity they're allowed to access.
* Add analytics events for when people see this new upsell, and for when they click its call-to-action.
* Modify `ActivityCardList` to only show events within the bounds of the current site's retention policy, if the feature flag is enabled.
* Modify the Activity Log to reflect a loading state until the current site's retention policy is known.

#### Known issues

* The button/link in the call-to-action currently redirects to `/pricing`, because a context-appropriate upsell page doesn't yet exist.
* If an error occurs while fetching the site's retention policy, the Activity Log will show a loading state forever.
* This implementation only covers Calypso Green. We will presumably need a similar pull request to cover Calypso Blue, borrowing functionality seen here.

#### Testing instructions

**IMPORTANT:** To test this pull request, you must configure your testing environment so that it accesses a development sandbox version of the WordPress.com API, and that sandbox must have `D64586-code` applied to it. Otherwise, you'll only see a loading state on the Activity Log page after the request to get your site's retention policy fails.

##### Setup

1. Before starting Calypso Green, change the value of the `activity-log/retention-policies` feature flag to true for your environment; or, add `?flags=activity-log/retention-policies` to your browser's URL as you test.
2. Select a site with events in its Activity Log, and visit its Activity Log.

##### Upsell visible

1. If your site has more than 30 days of events, scroll that far back in the Activity Log. Verify you see an upsell prompt on the last page, as shown in the reference screenshots.
2. Manually change your site's retention policy in the application state by running (e.g.) `state.activityLog.retentionPolicy[ <your_site_id> ].days = 1;` in the development console. Click on the filter bar at the top of the page, then dismiss it to cause the list of Activities to re-render.
3. Verify the list displays appropriately, hiding activities that would fall outside the retention period you set.
4. Verify that the upsell is visible and references the number of days you set in step 2.

##### Upsell not visible: flag disabled

1. Load the Activity Log for a site where you would normally see the retention limit upsell (see test cases above).
2. Add `?flags=-activity-log/retention-policies` to your browser's URL and hit Enter/Return to reload the page (note the minus sign).
3. Verify the retention limit upsell is no longer visible, and activities that would fall outside the retention policy are no longer hidden.

##### Upsell not visible: not enough activities

1. Load the Activity Log for a site with 30 days of events or fewer; alternatively, increase the retention period by modifying the application state as described in the "Upsell visible" case above.
2. Verify that you do not see the retention limit upsell described in the previous case, even scrolling to the last available page.

##### Analytics / tracking

1. Load the Activity Log for a site where you would normally be able to see the retention limit upsell (see test cases above), but _do not_ scroll it into view yet.
2. Open the Network panel of your browser's development console so you can see outgoing requests to `t.gif`.
3. Watching the Network panel as you do so, scroll the upsell into view.
4. Verify that only once the "Upgrade storage" button is fully in view, a request is sent with the event name `calypso_activitylog_retentionlimit_view`. A `site_id` parameter should be also included in the request.
5. Scroll back up the page until the upsell is no longer in view, then scroll back down. Verify the request isn't sent a second time.
6. Click the "Upgrade storage" button, verifying that another request is sent with the event name `calypso_activitylog_retentionlimit_upgrade_click` and a `site_id` parameter.

#### Reference screenshots

<img height="350" alt="Screen Shot 2021-07-26 at 10 52 52" src="https://user-images.githubusercontent.com/670067/127019938-14fc6657-dfe7-4e59-88f6-52838583949a.png"> <img height="350" alt="Screen Shot 2021-07-26 at 10 52 31" src="https://user-images.githubusercontent.com/670067/127019935-e8311e9c-7824-4f19-9046-e31fb75bf60b.png">
